src/Modules/Visualization/InterfaceWithOspray.cc

```diff
@@ -89,50 +98,137 @@ void InterfaceWithOspray::setStateDefaults()
+    OSPFrameBuffer framebuffer_;
+
+    float toFloat(const Name& name) const
     {
-      auto facade(field->mesh()->getFacade());
+      return static_cast<float>(state_->getValue(name).toDouble());
+    }
 
-      // image size
-      osp::vec2i imgSize;
-      imgSize.x = state->getValue(Parameters::ImageWidth).toInt();
-      imgSize.y = state->getValue(Parameters::ImageHeight).toInt();
+    struct FieldData
+    {
+      std::vector<float> vertex, color;
+      std::vector<int32_t> index;
+    };
 
-      auto toFloat = [state](const Name& name) { return static_cast<float>(state->getValue(name).toDouble()); };
+    std::vector<FieldData> fieldData_;
+
+  public:
+    explicit OsprayImpl(ModuleStateHandle state) : guard_(lock_.get()), state_(state)
+    {
+      initialize();
+    }
+
+    void setup()
+    {
+      imgSize_.x = state_->getValue(Parameters::ImageWidth).toInt();
+      imgSize_.y = state_->getValue(Parameters::ImageHeight).toInt();
 
       // camera
       float cam_pos[] = { toFloat(Parameters::CameraPositionX), toFloat(Parameters::CameraPositionY), toFloat(Parameters::CameraPositionZ) };
       float cam_up[] = { toFloat(Parameters::CameraUpX), toFloat(Parameters::CameraUpY), toFloat(Parameters::CameraUpZ) };
       float cam_view[] = { toFloat(Parameters::CameraViewX), toFloat(Parameters::CameraViewY), toFloat(Parameters::CameraViewZ) };
 
-      auto map = colorMap.value_or(nullptr);
-      std::vector<float> vertex, color;
+      // create and setup camera
+      camera_ = ospNewCamera("perspective");
+      ospSetf(camera_, "aspect", imgSize_.x / (float)imgSize_.y);
+      ospSet3fv(camera_, "pos", cam_pos);
+      ospSet3fv(camera_, "dir", cam_view);
+      ospSet3fv(camera_, "up", cam_up);
+      ospCommit(camera_); // commit each object to indicate modifications are done
+
+      world_ = ospNewModel();
+      ospCommit(world_);
+    }
+
+    void adjustCameraPosition(FieldHandle field)
+    {
+      if (state_->getValue(Parameters::AutoCameraView).toBool())
+      {
+        auto vmesh = field->vmesh();
+        auto bbox = vmesh->get_bounding_box();
+        imageBox_.extend(bbox);
+        auto center = imageBox_.center();
+        float position[] = { toFloat(Parameters::CameraPositionX), toFloat(Parameters::CameraPositionY), toFloat(Parameters::CameraPositionZ) };
+        float newDir[] = { static_cast<float>(center.x()) - position[0],
+           static_cast<float>(center.y()) - position[1],
+           static_cast<float>(center.z()) - position[2]};
+        //std::cout << "newDir " << newDir[0] << ", " << newDir[1] << ", " << newDir[2] << std::endl;
+        state_->setValue(Parameters::CameraViewX, center.x());
+        state_->setValue(Parameters::CameraViewY, center.y());
+        state_->setValue(Parameters::CameraViewZ, center.z());
+        ospSet3fv(camera_, "dir", newDir);
+        float newUp[] = { newDir[0] / newDir[2], -(newDir[0]*newDir[0] + newDir[2]*newDir[2])/(newDir[1]*newDir[2]) , 1.0f };
```

Comment:
This would be better if this was something like:
newUp=cross(cross(newDir,cam_up),newDir),eg,
float side[] = {newDir[1]*cam_up[2] - newDir[2]*cam_up[1],newDir[2]*cam_up[0] - newDir[0]*cam_up[2], newDir[0]*cam_up[1] - newDir[1]*cam_up[0], 1.0f };
float norm_side = sqrt(side[0]*side[0]+side[1]*side[1]+side[2]*side[2]);
if (norm_side <= 1e-3) 
{
side[] = {newDir[1] ,- newDir[0], 0.0f, 1.0f };
norm_side = sqrt(side[0]*side[0]+side[1]*side[1]+side[2]*side[2]);
if (norm_side <= 1e-3) 
{ float side[] = {- newDir[2],0.0f, newDir[0], 1.0f };
norm_side = sqrt(side[0]*side[0]+side[1]*side[1]+side[2]*side[2]);
}
}
side = side/norm_side;
float newUp[] = {side[1]*newDir[2] - side[2]*newDir[1],side[2]*newDir[0] - side[0]*newDir[2],side[0]*newDir[1] - side[1]*newDir[0],1.0f}

Reply:
This would try to preserve something similar to the original up vector, if it is too close, it would use z axis, and the third choice is y axis. 

The current version, can twist the camera in a strange way.